### ISSUE TITLE
Systemd and volumio files not properly updated

### ISSUE THREAD

**Issue by volumio** [volumio/Build#479]
@ashthespy 

Issue:
- The custom files in https://github.com/volumio/Build/tree/volumioOS/volumio are copied over too soon.
- When installing custom packages https://github.com/volumio/Build/blob/volumioOS/build.sh#L548 for example, the systemd file in the custom package overwrites the volumio custom one

GOAL:
- Have the files  https://github.com/volumio/Build/tree/volumioOS/volumio copied as late as possible in the build process, following the pattern used in https://github.com/volumio/Build/blob/master/scripts/configure.sh
- This will make sure that all the customization we've done on those files are respected
- Please send a PR (tested) to volumioOS

**Comment by ashthespy** [comment]
This is something I wanted to clarify -- this part is from the old build system that I copied over, but it would be good to refactor.


#### Rootfs Creation
-- Run `multistrap` ([volumioOS](https://github.com/volumio/Build/blob/7e70a9a70ccd64d4428d7e4fd88b97d5d2d1a8ba/build.sh#L367)  [master](https://github.com/volumio/Build/blob/e81b2037a65d8f554cb0c857a1007d659ad7f838/build.sh#L162))
-- Run `volumioconfig.sh` in chroot ([volumioOS](https://github.com/volumio/Build/blob/7e70a9a70ccd64d4428d7e4fd88b97d5d2d1a8ba/build.sh#L395) [master](https://github.com/volumio/Build/blob/master/build.sh#L201))
-- Run `configure.sh` on the resulting rootfs ([volumioOS](https://github.com/volumio/Build/blob/7e70a9a70ccd64d4428d7e4fd88b97d5d2d1a8ba/build.sh#L419) [master](https://github.com/volumio/Build/blob/master/build.sh#L227))


#### Image creation
 -- Run [`<device>image.sh`](https://github.com/volumio/Build/blob/e81b2037a65d8f554cb0c857a1007d659ad7f838/build.sh#L240) ([`makeimage.sh`](https://github.com/volumio/Build/blob/7e70a9a70ccd64d4428d7e4fd88b97d5d2d1a8ba/build.sh#L561) in volumioOS
 --- Run [`<device>config.sh`](https://github.com/volumio/Build/blob/e81b2037a65d8f554cb0c857a1007d659ad7f838/scripts/raspberryimage.sh#L109) in chroot. ([`chrootconfig.sh`](https://github.com/volumio/Build/blob/7e70a9a70ccd64d4428d7e4fd88b97d5d2d1a8ba/scripts/makeimage.sh#L228) in volumioOS)
 --- Run `finalize.sh`([volumioOS](https://github.com/volumio/Build/blob/7e70a9a70ccd64d4428d7e4fd88b97d5d2d1a8ba/scripts/makeimage.sh#L241) [master](https://github.com/volumio/Build/blob/e81b2037a65d8f554cb0c857a1007d659ad7f838/scripts/raspberryimage.sh#L134))

The structure here is a bit tricky, as `volumioconfig` seems to have initially grown from [davidmonro/pi_builder/configscript.sh](https://github.com/davidmonro/pi_builder/blob/master/configscript.sh) (or the other way around?) -- but the purpose is mainly to finish the multistrap Debconf and pre-seeding. 
But as `volumioconfig` is currently, it is doing some sketchy things like symlinking non existent files -- the service files are copied over only later in `configure.sh`.  It also looks like some tried enabling the service, and then found a workaround.. ;-)
https://github.com/volumio/Build/blob/e81b2037a65d8f554cb0c857a1007d659ad7f838/scripts/volumioconfig.sh#L461-L463


#### Which leads to the question(s)
##### -- Should `volumioconfig` be trimmed back to it's original purpose -- finish Debconf configuration after multistrap? 
##### -- When should which configuration be actually done? 
 --- Some configuration files are created in `volumioconfig.sh` 
https://github.com/volumio/Build/blob/7e70a9a70ccd64d4428d7e4fd88b97d5d2d1a8ba/scripts/volumio/volumioconfig.sh#L361-L366
 --- Some files from `/volumio` are copied over in `configure.sh`
https://github.com/volumio/Build/blob/7e70a9a70ccd64d4428d7e4fd88b97d5d2d1a8ba/scripts/volumio/configure.sh#L67-L68

tl;dr --- We should consider refactoring this part of the build system as well - I only worked on the modularity of the device creation, but it would be good to go consider what is actually required, and how to do it. 

That being said, I am a bit preoccupied with life until end of March, so might be good to find a short term solution if this is a blocker.

**Comment by volumio** [comment]
@ashthespy IMHO, given the lack of time and the absolute priority (which is sending a testable build to the community ASAP) I suggest:
- Working on the immediate issue, let's keep refactors for later
- Originally, custom packages were installed in volumioconfig and THEN configure would place the correct files in the system (including systemd )
https://github.com/volumio/Build/blob/master/build.sh#L209
- Let's just add "configure" after the install of custom packages, or move the install of custom packages before configure (or whenever the files are copied)

That will unlock the situation and allow to start testing ;) After that your ideas of refactoring can be good.

**Comment by ashthespy** [comment]
>    * Originally, custom packages were installed in volumioconfig and THEN configure would place the correct files in the system (including systemd )
>    * Let's just add "configure" after the install of custom packages, or move the install of custom packages before configure (or whenever the files are copied)

Yep, you could still do that - have a look at how the configuration variables are passed into the chroot scripts and packages are installed  [`chrootconfig.sh`](https://github.com/volumio/Build/blob/7e70a9a70ccd64d4428d7e4fd88b97d5d2d1a8ba/scripts/volumio/chrootconfig.sh#L62-L68) you could copy that over to `volumioconfig` so that is done at a rootfs level, rather than at the device level.

**Comment by volumio** [comment]
Thanks for the hint. What I still don't understand is when the custom systemd files are copied.
Could you point me to that please?

**Comment by ashthespy** [comment]
That part remains unchanged - it will be the same as in the old system so should be somewhere in `configure.sh`

**Comment by volumio** [comment]
@ashthespy I tried with
https://github.com/volumio/Build/commit/23d70acea9c27e2819838a3be7f0378d726c0c2a

but packages do not get installed... 
Sorry if I ask, but I need pointers here since I still struggle to get the chronology of what is called when.
When would you place the custom package install so it gets installed before configure?

**Comment by ashthespy** [comment]
Hmm with 23d70ac you would break the [`KIOSKMODE`](https://github.com/volumio/Build/blob/1c81292089a777ca989ffeb8228217a96fcf0ffa/recipes/devices/families/x86.sh#L24) which is a device level configuration. I don't believe you want ALL rootfs's created to contain all the kiosk dependencies by default?

>  When would you place the custom package install so it gets installed before configure?

-- For that snippet to work,  `/volumio/customPkgs` needs to contain the correct packages.  So either they would have to be extracted into rootfs either by multistrap directly, or fetched from the repo in `volumioconfig` as previously, or copied into the rootfs manually before `volumioconfig.sh` is run..

**Comment by ashthespy** [comment]
Issues after switching to Volumio builds of  `shairport` and `upmpdcli`  are  fixed with https://github.com/volumio/Build/commit/e446a7f1b7b570d35771f25d01639d8fc4166f4a 

Leaving this open for future refactoring..
